p2p/src/secret_connection.rs

```diff
@@ -148,19 +168,79 @@ impl<IoHandler: Read + Write + Send + Sync> SecretConnection<IoHandler> {
         let remote_sig = ed25519::Signature::try_from(auth_sig_msg.sig.as_slice())
             .map_err(|_| Error::CryptoError)?;
 
-        if protocol_version.has_transcript() {
+        if self.protocol_version.has_transcript() {
             remote_pubkey
-                .verify(&sc_mac, &remote_sig)
+                .verify(&self.state.sc_mac, &remote_sig)
                 .map_err(|_| Error::CryptoError)?;
         } else {
             remote_pubkey
-                .verify(&kdf.challenge, &remote_sig)
+                .verify(&self.state.kdf.challenge, &remote_sig)
                 .map_err(|_| Error::CryptoError)?;
         }
 
         // We've authorized.
-        sc.remote_pubkey = Some(remote_pubkey.into());
+        Ok(remote_pubkey.into())
+    }
+}
+
+/// Encrypted connection between peers in a Tendermint network
+pub struct SecretConnection<IoHandler: Read + Write + Send + Sync> {
+    io_handler: IoHandler,
+    protocol_version: Version,
+    recv_nonce: Nonce,
+    send_nonce: Nonce,
+    recv_cipher: ChaCha20Poly1305,
+    send_cipher: ChaCha20Poly1305,
+    remote_pubkey: Option<PublicKey>,
+    recv_buffer: Vec<u8>,
+}
+
+impl<IoHandler: Read + Write + Send + Sync> SecretConnection<IoHandler> {
+    /// Returns authenticated remote pubkey
+    pub fn remote_pubkey(&self) -> PublicKey {
+        self.remote_pubkey.expect("remote_pubkey uninitialized")
+    }
+
+    /// Performs handshake and returns a new authenticated SecretConnection.
+    pub fn new(
+        mut io_handler: IoHandler,
+        local_privkey: ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<SecretConnection<IoHandler>> {
+        // Start a handshake process.
+        let local_pubkey = PublicKey::from(&local_privkey);
+        let (mut h, local_eph_pubkey) = Handshake::new(local_privkey, protocol_version);
 
+        // Write local ephemeral pubkey and receive one too.
+        let remote_eph_pubkey =
+            share_eph_pubkey(&mut io_handler, &local_eph_pubkey, protocol_version)?;
+
+        // Compute a local signature (also recv_cipher & send_cipher)
+        let mut h = h.got_key(remote_eph_pubkey)?;
+
+        let mut sc = SecretConnection {
+            io_handler,
+            protocol_version,
+            recv_buffer: vec![],
+            recv_nonce: Nonce::default(),
+            send_nonce: Nonce::default(),
+            recv_cipher: h.state.recv_cipher.clone(),
```

Comment:
note: I had to clone ciphers. Otherwise, I'd have to move them out of the Handshake struct.

Reply:
It turns out it's not so easy. In addition to the ciphers, there is also a nonce, which is incremented every time we send a packet with encrypted data. So I would need to extract 4 methods (encrypt, decrypt, read, write; read/write also depend on buffers), which doesn't seem like such a good idea to me.

Reply:
Sensible.